Scenes: Add snapshots view support

evictorero opened the pull request:
**What is this feature?**

Adding Scenes support for rendering a Snapshot created using Scenes.
At this moment, old Snapshots won't work. We would need additional work.
![Screen Recording 2024-01-30 at 16 24 40](https://github.com/grafana/grafana/assets/20256983/ab734c2b-33a7-41fa-a297-30b92baf2038)



Fixes https://github.com/grafana/grafana/issues/81582

**Special notes for your reviewer:**
I separated the Snapshots code instead of adding special cases inside `loadingDashboard` and `loadScene` functions. 
I want to confirm this approach before writing tests.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

evictorero committed:
Scenes: Add snapshots view support

axelavargas reviewed (commented):
Nice progress 👏🏾 , couple of thoughts:

- I think we might need to discuss with UX where the "original dashboard" link should be placed 🤔 

![image](https://github.com/grafana/grafana/assets/239999/3032f6ef-ff43-4134-8e46-197261e9ba1a)

- Somehow rendering the snapshot in scenes, is giving me annotations (it does not happen in core Grafana)

![image](https://github.com/grafana/grafana/assets/239999/56248707-18f7-405d-9b04-b25c07d0094a)

evictorero committed:
splitted code for snapshots and enabled view as scene button in original snapshot

evictorero committed:
remove annotations call and add view as dashbaord on snapshot scenes view

evictorero committed:
remove unused property

evictorero committed:
add open original dashboard button

evictorero commented:
> I think we might need to discuss with UX where the "original dashboard" link should be placed 🤔

Added in the `DashNav` after UX feedback

> Somehow rendering the snapshot in scenes, is giving me annotations (it does not happen in core Grafana)

Removed this request since the snapshots annotations will be inside the snapshot data. You probably tested logged in but this was raising an auth error when seeing without authentication because the annotations request is authenticated.

evictorero commented on public/app/features/dashboard-scene/scene/DashboardScene.tsx:
```diff
@@ -388,6 +388,16 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
 
+  onOpenSnapshotOriginalDashboard = () => {
+    try {
+      // @ts-ignore
```
Should I add this `originalUrl` in the schema files?

evictorero commented on public/app/features/dashboard/components/DashNav/DashNav.tsx:
```diff
@@ -192,7 +192,7 @@ export const DashNav = React.memo<Props>((props) => {
     }
 
-    if (config.featureToggles.scenes && !dashboard.isSnapshot()) {
+    if (config.featureToggles.scenes) {
```
We want to show now the button to render a snapshot as a Scene.

evictorero commented on public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts:
```diff
@@ -100,6 +100,27 @@ export class DashboardScenePageStateManager extends StateManagerBase<DashboardSc
     return rsp;
   }
 
+  public async loadSnapshot(slug: string) {
```
These logic could be inside `loadDashboard` and `loadScene` but I didn't want to introduce additional logic related to Snapshots. I could change it if it does not make sense.

axelavargas commented on public/app/features/dashboard-scene/scene/DashboardScene.tsx:
```diff
@@ -388,6 +388,16 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
     }
   }
 
+  onOpenSnapshotOriginalDashboard = () => {
```
I think we could reuse the function from DashNav
https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/components/DashNav/DashNav.tsx#L88, there are some security checks we need :)

evictorero replied:
Yes, I saw them and tried to reproduce the opening of the modal and I couldn't. There is a condition that never will be true `sanitizedUrl.host !== appUrl.host` because the `sanitizedUrl` is made with the same host `config.appUrl`. I think this is an old code that does not work anymore but I will need to dig more into this.

evictorero replied:
Good catch! I've added these checks to the new code.

axelavargas reviewed (commented):
Hey @evictorero  😄, love the progress with this PR 🏅 

I think the approach of separating the logic of snapshots is a good one, so feel free to add the test.

P.S. I left some comments regarding the 'original dashboard' link.

evictorero committed:
add tests and security checks on opening original dashboard

evictorero committed:
Merge branch 'main' into evictorero/scenes-snapshot-support

axelavargas reviewed (approved):
Hey @evictorero 👋🏾, I checked the code, and it looks great (🏅 nice tests).  

I also gave it a try locally with a couple of snapshots that are using the format and it works as expected.

torkelo commented on public/app/features/dashboard-scene/scene/NavToolbarActions.tsx:
```diff
@@ -126,6 +136,19 @@ export const NavToolbarActions = React.memo<Props>(({ dashboard }) => {
+
+    if (meta.isSnapshot) {
+      toolbarActions.push(
+        <DashNavButton
+          key="button-snapshot"
+          tooltip={t('dashboard.toolbar.open-original', 'Open original dashboard')}
+          icon="link"
+          onClick={() => {
+            dashboard.onOpenSnapshotOriginalDashboard();
+          }}
+        />
```
I would put this in a new GoToSnapshotOriginButton react component and put the logic of onOpenSnapshotOriginalDashboard inside it (to not pollute DashboardScene with it)

evictorero replied:
I will do it in another PR so I avoid future conflicts, these files are changing heavily.

evictorero committed:
Merge branch 'main' into evictorero/scenes-snapshot-support

evictorero committed:
fix test

evictorero committed:
improve toolbar buttons

evictorero committed:
skip fetch dashboard when rendering a snapshot

evictorero committed:
removing unwanted changes
